lib/elixir/lib/calendar/datetime.ex

```diff
@@ -177,20 +177,209 @@ defmodule DateTime do
   Converts the given `NaiveDateTime` to `DateTime`.
 
   It expects a time zone to put the NaiveDateTime in.
-  Currently it only supports "Etc/UTC" as time zone.
+
+  It only supports "Etc/UTC" as time zone if a `TimeZoneDatabase`
+  is not provided as a third argument.
 
   ## Examples
 
       iex> {:ok, datetime} = DateTime.from_naive(~N[2016-05-24 13:26:08.003], "Etc/UTC")
       iex> datetime
       #DateTime<2016-05-24 13:26:08.003Z>
 
+  When the datetime is ambiguous - for instance during changing from summer
+  to winter time - the two possible valid datetimes are returned. First the one
+  that happens first, then the one that happens after.
+
+      iex> {:ambiguous, first_dt, second_dt} = DateTime.from_naive(~N[2018-10-28 02:30:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> first_dt
+      #DateTime<2018-10-28 02:30:00+02:00 CEST Europe/Copenhagen>
+      iex> second_dt
+      #DateTime<2018-10-28 02:30:00+01:00 CET Europe/Copenhagen>
+
+  When there is a gap in wall time - for instance in spring when the clocks are
+  turned forward - the latest valid datetime just before the gap and the first
+  valid datetime just after the gap.
+
+      iex> {:gap, just_before, just_after} = DateTime.from_naive(~N[2019-03-31 02:30:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> just_before
+      #DateTime<2019-03-31 01:59:59.999999+01:00 CET Europe/Copenhagen>
+      iex> just_after
+      #DateTime<2019-03-31 03:00:00+02:00 CEST Europe/Copenhagen>
+
+  Most of the time there is one, and just one, valid datetime for a certain
+  date and time in a certain time zone.
+
+      iex> {:ok, datetime} = DateTime.from_naive(~N[2018-07-28 12:30:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> datetime
+      #DateTime<2018-07-28 12:30:00+02:00 CEST Europe/Copenhagen>
+
   """
   @doc since: "1.4.0"
-  @spec from_naive(NaiveDateTime.t(), Calendar.time_zone()) :: {:ok, t}
-  def from_naive(naive_datetime, time_zone)
+  @spec from_naive(
+          NaiveDateTime.t(),
+          Calendar.time_zone(),
+          TimeZoneDatabaseClient.tz_db_or_config()
+        ) ::
+          {:ok, t}
+          | {:outside_leap_second_data_range, t}
+          | {:ambiguous, t, t}
+          | {:gap, t, t}
+          | {:error, :time_zone_not_found}
+          | {:error, :incompatible_calendars}
+          | {:error, :no_time_zone_database}
+
+  def from_naive(naive_datetime, time_zone, tz_db_or_config \\ :from_config)
+
+  def from_naive(%{second: 60} = naive_datetime, "Etc/UTC", tz_db_or_config) do
```

Comment:
This is specific to the ISO calendar. I think we will have to move the leap seconds handling to the ISO calendar but I am not quite sure about how to do that.

Reply:
Stream of thoughts:

The fact we may have negative leap seconds means that we also need to check for `%{second: 59}` and see if it actually exists. I understand that we haven't had a negative leap second yet but not handling those cases (or not thinking explicitly about them) may mean two things:

1. whenever they happen, old elixir versions will be unable to cope with them
2. our API may be unable to be augmented whenever they are introduced

My suggestion is for leap seconds to be layered into DateTime somehow. This can either be `DateTimeWithLeapSecond` (or a new field in DateTime) or a `Calendar.ISOWithLeapSecond`. The latter would mean that `Calendar.ISO` today should have been called `Calendar.ISOWithoutLeapSecond` but unfortunately this ship has sailed. I will have to think more about this but this is inlined to other libraries where leap seconds are opted in.

Reply:
> The fact we may have negative leap seconds

From what I have read in various places it is unlikely that a negative leap second will occur anytime soon.

From https://web.archive.org/web/20130602095155/http://iraf.noao.edu/~seaman/leap/leapsec.txt :

> One thing to note is the steep slope resulting from initializing UT to the
 old Ephemeris Time definition of so many seconds in 1900.  This slope all
 but guarantees that a negative leap second will never occur - not only would
 the Earth's current slowing trend have to be combatted, but the accumulated
 bias over a coherent 1.5 year timescale would have to be reversed.  We don't
 have leap seconds because of current slowing - we have leap seconds because
 the Earth has already spun down.

Reply:
Awesome, thanks for the info!